component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java

```diff
@@ -72,185 +73,322 @@
-        String jti = claimsSet.getJWTID();
-        long currentTimeInMillis = System.currentTimeMillis();
-        long timeStampSkewMillis = OAuthServerConfiguration.getInstance().getTimeStampSkewInSeconds() * 1000;
-
-        if (isEmpty(jwtIssuer) || isEmpty(jwtSubject) || expirationTime == null || audience ==
-                null || (preventTokenReuse && jti == null)) {
-            return logAndReturnFalse("Mandatory fields(Issuer, Subject, Expiration time , " +
-                    "JWT ID or Audience) are empty in the given JSON Web Token.");
-        }
+            if (claimsSet == null) {
+                errorMessage = "Claim set is missing in the JWT assertion";
+                if (log.isDebugEnabled()) {
+                    log.debug(errorMessage);
+                }
+                throw new OAuthClientAuthnException(errorMessage, OAuth2ErrorCodes.INVALID_REQUEST);
+            }
 
-        if (isNotEmpty(validIssuer) && !validIssuer.equals(jwtIssuer)) {
-            return logAndReturnFalse("Invalid Issuer:" + jwtIssuer + " in the given JSON Web Token.");
+            String jwtIssuer = claimsSet.getIssuer();
+            String jwtSubject = resolveSubject(claimsSet);
+            List<String> audience = claimsSet.getAudience();
+            Date expirationTime = claimsSet.getExpirationTime();
+            String jti = claimsSet.getJWTID();
+            Date nbf = claimsSet.getNotBeforeTime();
+            Date issuedAtTime = claimsSet.getIssueTime();
+            long currentTimeInMillis = System.currentTimeMillis();
+            long timeStampSkewMillis = OAuthServerConfiguration.getInstance().getTimeStampSkewInSeconds() * 1000;
+            OAuthAppDO oAuthAppDO = getOAuthAppDO(jwtSubject);
+            String tenantDomain = oAuthAppDO.getUser().getTenantDomain();
+
+            if (!isValidSignature(signedJWT, tenantDomain, jwtSubject)) {
+                return false;
+            }
+            //Validate issuer and subject.
+            if (!validateIssuer(jwtIssuer, oAuthAppDO) || !validateSubject(jwtSubject)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Invalid jwtIssuer: " + jwtIssuer + " or jwtSubject: " + jwtSubject + " is found in the assertion." +
+                            " Expected value for issuer and subject is: " + oAuthAppDO.getOauthConsumerKey());
+                }
+                return false;
+            }
+
+            //Get audience.
+            String validAud = getValidAudience(tenantDomain);
+            long expTime = 0;
+            long issuedTime = 0;
+            if (expirationTime != null) {
+                expTime = expirationTime.getTime();
+            }
+            if (issuedAtTime != null) {
+                issuedTime = issuedAtTime.getTime();
+            }
+
+            //Validate mandotory feilds, audience, nbf,exp time, jti.
+            if (!validateMandatoryFeilds(mandatoryClaims, claimsSet) || !validateAudience(validAud, audience) ||
```

Comment:
can move validating mandatory fields before validating signature and issuer. If mandatory fields are not present in JWT , no need to proceed.